### PR TITLE
Apt work

### PR DESCRIPTION
**Pull request by feliwir**
Things done in this PR:
- Implemented all parts of the Actionscript VM required for running the BFME II Mainmenu
- Changed InstructionReader to InstructionCollection for storing & parsing bytecode

Things still required:
- Handle buttons
- Read the text variables from actionscript and localize the strings
- Implement missing Instrutions for other .apts
- handle missing FrameItems & Characters

### CODE REVIEW

File: src/OpenSage.Game/Gui/Apt/ActionScript/ActionContext.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        public ObjectContext Scope { get; set; }
+        public ObjectContext Global { get; set; }
+        public AptContext Apt { get; set; }
+        public InstructionStream Stream { get; set; }
+        public Stack<Value> Stack { get; set; }
+        public Value[] Registers { get; set; }
+        public Dictionary<string, Value> Params { get; set; }
+
+
+        public ActionContext(int reg = 0)
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/ActionContext.cs]: Can this be renamed to `numRegisters` or `registerCount` please?

File: src/OpenSage.Game/Gui/Apt/ActionScript/InstructionStream.cs [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        /// </summary>
+        /// <returns></returns>
+        public InstructionBase GetInstruction()
+        {
+            if (_position - 1 > _instructions.Count)
+                throw new IndexOutOfRangeException();
+
+            //skip any possible padding
+            if (_instructions[_position].Type == InstructionType.Padding)
+                ++_position;
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/InstructionStream.cs]: Can you add braces around this line please.

File: src/OpenSage.Game/Gui/Apt/ActionScript/InstructionStream.cs [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        }
+
+        /// <summary>
+        /// Get an instruction and move to the next instruction right away. Skip paddings
+        /// </summary>
+        /// <returns></returns>
+        public InstructionBase GetInstruction()
+        {
+            if (_position - 1 > _instructions.Count)
+                throw new IndexOutOfRangeException();
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/InstructionStream.cs]: Can you add braces around this line please.

File: src/OpenSage.Game/Gui/Apt/ActionScript/InstructionStream.cs [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        private uint CalculateByteOffset(int instr)
+        {
+            uint size = 0;
+
+            for (int i = 0; i < instr; ++i)
+            {
+                size += _instructions[i].Size;
+
+                if (_instructions[i].Type != InstructionType.Padding)
+                    ++size;
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/InstructionStream.cs]: Missing braces.

File: src/OpenSage.Game/Gui/Apt/ActionScript/ObjectContext.cs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        /// <param name="value">value name</param>
+        /// <returns></returns>
+        public Value ResolveValue(string value)
+        {
+            var path = value.Split('.');
+            var obj = this;
+
+            if(path.Length>1)
+            {
+                if (path[0] == "_parent")
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/ActionScript/ObjectContext.cs]: How does this logic relate to the check for "_parent" in `ActionContext.GetValue`?

File: src/OpenSage.Game/Gui/Apt/DisplayItem.cs [hunk 6]
```diff
@@ -1,26 +1,30 @@
...
-        public static readonly ItemTransform None = new ItemTransform(ColorRgbaF.White, Matrix3x2.Identity);
+        public static readonly ItemTransform None = new ItemTransform(ColorRgbaF.White, Matrix3x2.Identity,Vector2.Zero);
 
         public ColorRgbaF ColorTransform;
-        public Matrix3x2 GeometryTransform;
+        public Matrix3x2 GeometryRotation;
+        public Vector2 GeometryTranslation;
 
-        public ItemTransform(ColorRgbaF color, Matrix3x2 geometry)
+        public ItemTransform(ColorRgbaF color, Matrix3x2 rotation,Vector2 translation)
```
Comment (tgjones) [inline src/OpenSage.Game/Gui/Apt/DisplayItem.cs]: Missing a space :)

File: src/OpenSage.Game/Gui/Apt/ActionScript/ObjectContext.cs (hunk 5 above)
Comment (feliwir) [reply src/OpenSage.Game/Gui/Apt/ActionScript/ObjectContext.cs]: In ObjectContext it is asked for a specific variable relative to the current object, e.g. "_parent.buttonColor" or just a variable inside the current object "variable" etc. In ActionContext.GetValue it is checked for builtin variables in general. It's not quite nice yet, but i'll hopefully be able to clean the system up

**Comment by tgjones** [comment]
Thanks!

**Pull request merged**